File: app/services/agent_session_parsers/codex.rb
module AgentSessionParsers
  class Codex < Base
    def parse
      records = parse_jsonl_lines
      messages = []
      current_assistant_blocks = []

      records.each do |record|
        case record["type"]
        when "event_msg"
          handle_event_msg(record, messages, current_assistant_blocks)
        when "response_item"
          handle_response_item(record, messages, current_assistant_blocks)
        when "item.completed"
          handle_item_completed(record, messages, current_assistant_blocks)
        when "turn.completed", "turn_context"
          flush_assistant_blocks(messages, current_assistant_blocks)
        end
      end

      flush_assistant_blocks(messages, current_assistant_blocks)

      metadata = extract_metadata(records, messages)
      build_result(messages: messages, metadata: metadata)
    end

    private

    def handle_event_msg(record, messages, current_assistant_blocks)
      payload = record["payload"] || {}
      return unless payload["type"] == "user_message"

      flush_assistant_blocks(messages, current_assistant_blocks)
      text = payload["message"] || payload["text"] || payload.dig("content", 0, "text") || ""
      return if text.blank?

      messages << build_message(role: "user", content_blocks: [text_block(text)])
    end

    # New Codex format (0.105+): response_item records
    def handle_response_item(record, messages, current_assistant_blocks)
      payload = record["payload"] || {}
      role = payload["role"]
      item_type = payload["type"]

      case item_type
      when "message"
        if role == "user"
          # In new Codex format, user messages come from event_msg (user_message payload).
          # response_item role=user contains duplicates plus system instructions — skip them.
          return
        elsif role == "assistant"
          text = extract_response_item_text(payload)
          if text.present?
            # Flush any pending tool calls before starting text
            flush_assistant_blocks(messages, current_assistant_blocks) if current_assistant_blocks.any? { |b| b["type"] == "tool_call" }
            current_assistant_blocks << text_block(text)
          end
        end
        # Skip developer role (system instructions)
      when "function_call", "custom_tool_call"
        # Flush any pending text blocks as their own message
        flush_assistant_blocks(messages, current_assistant_blocks)

        name = payload["name"] || payload["call_id"] || "tool_call"
        input = payload["arguments"] || payload["input"]
        # Each tool call becomes its own message (output paired later)
        current_assistant_blocks << tool_call_block(
          name: name,
          input: input.is_a?(String) ? input.truncate(200) : input&.to_json&.truncate(200),
          output: nil,
        )
        flush_assistant_blocks(messages, current_assistant_blocks)
      when "function_call_output", "custom_tool_call_output"
        output = extract_output_content(payload)
        formatted = truncate_output(output)
        # Find the first tool_call in already-flushed messages that has no output yet
        attach_output_to_first_unmatched(messages, formatted)
      when "reasoning"
        summary = payload.dig("summary", 0, "text")
        if summary.present?
          flush_assistant_blocks(messages, current_assistant_blocks) if current_assistant_blocks.any? { |b| b["type"] == "tool_call" }
          current_assistant_blocks << text_block(summary)
        end
      end
    end

    # Old Codex format: item.completed records
    def handle_item_completed(record, messages, current_assistant_blocks)
      item = record["item"] || {}
      item_type = item["type"]

      case item_type
      when "message"
        text = extract_message_text(item)
        if text.present?
          flush_assistant_blocks(messages, current_assistant_blocks) if current_assistant_blocks.any? { |b| b["type"] == "tool_call" }
          current_assistant_blocks << text_block(text)
        end
      when "function_call", "command"
        flush_assistant_blocks(messages, current_assistant_blocks) if current_assistant_blocks.any? { |b| b["type"] == "text" }
        name = item["name"] || item["command"] || "command"
        input = item.dig("arguments") || item.dig("input") || item.dig("command")
        raw_output = item.dig("output") || item.dig("result")
        output_text = unwrap_json_output(raw_output)
        current_assistant_blocks << tool_call_block(
          name: name,
          input: input.is_a?(String) ? input : input&.to_json&.truncate(200),
          output: truncate_output(output_text.is_a?(String) ? output_text : output_text&.to_json),
        )
        flush_assistant_blocks(messages, current_assistant_blocks)
      when "file_change"
        flush_assistant_blocks(messages, current_assistant_blocks) if current_assistant_blocks.any? { |b| b["type"] == "text" }
        path = item["file_path"] || item["path"]
        current_assistant_blocks << tool_call_block(
          name: "FileChange",
          input: path,
          output: truncate_output(item["diff"] || item["content"]),
        )
        flush_assistant_blocks(messages, current_assistant_blocks)
      end
    end

    def extract_response_item_text(payload)
      content = payload["content"]
      case content
      when String then content
      when Array
        content.filter_map { |c|
          c["text"] if %w[output_text input_text text].include?(c["type"])
        }.join("\n")
      else
        payload["text"]
      end
    end

    def extract_message_text(item)
      content = item["content"]
      case content
      when String then content
      when Array
        content.filter_map { |c| c["text"] if c["type"] == "text" }.join("\n")
      else
        item["text"]
      end
    end

    # Extract meaningful text from tool output which may be a JSON wrapper.
    # Codex wraps some outputs as {"output":"...", "metadata":{...}} — either as Hash or JSON string.
    def extract_output_content(payload)
      raw = payload["output"] || payload["result"]
      unwrap_json_output(raw)
    end

    def unwrap_json_output(raw)
      case raw
      when Hash
        raw["output"].presence || raw["content"].presence || raw.to_json
      when String
        # Try to parse JSON strings that wrap the actual output
        if raw.strip.start_with?("{")
          begin
            parsed = JSON.parse(raw)
            return parsed["output"].presence || parsed["content"].presence || raw if parsed.is_a?(Hash)
          rescue JSON::ParserError
            # Not JSON, return as-is
          end
        end
        raw
      else
        raw&.to_json
      end
    end

    def attach_output_to_first_unmatched(messages, output)
      messages.each do |m|
        next unless m["role"] == "assistant"

        m["content"]&.each do |b|
          if b["type"] == "tool_call" && b["output"].nil?
            b["output"] = output
            return
          end
        end
      end
    end

    def flush_assistant_blocks(messages, blocks)
      return if blocks.empty?

      messages << build_message(role: "assistant", content_blocks: blocks.dup)
      blocks.clear
    end

    def extract_metadata(records, messages)
      # New format: session_meta
      session_meta = records.find { |r| r["type"] == "session_meta" }
      # Old format: thread.started
      started = records.find { |r| r["type"] == "thread.started" }

      meta_payload = session_meta&.dig("payload") || {}
      {
        "tool_name" => "codex",
        "session_id" => meta_payload["id"] || started&.dig("thread_id"),
        "start_time" => meta_payload["timestamp"] || session_meta&.dig("timestamp") || started&.dig("timestamp"),
        "model" => meta_payload.dig("model_provider"),
        "cli_version" => meta_payload["cli_version"],
        "total_messages" => messages.size,
      }.compact
    end
  end
end

File: app/javascript/agentSessionParsers/codex.js
import {
  parseJsonlLines, buildResult, buildMessage, textBlock, toolCallBlock,
  truncateOutput, truncateStr, attachOutputToToolCall,
} from './base.js';

const TEXT_CONTENT_TYPES = new Set(['output_text', 'input_text', 'text']);

export function parse(rawContent) {
  const records = parseJsonlLines(rawContent);
  const messages = [];
  let currentAssistantBlocks = [];

  const sessionModel = extractSessionModel(records);

  for (const record of records) {
    switch (record.type) {
      case 'event_msg':
        handleEventMsg(record, messages, currentAssistantBlocks, sessionModel);
        currentAssistantBlocks = [];
        break;
      case 'response_item':
        currentAssistantBlocks = handleResponseItem(record, messages, currentAssistantBlocks, sessionModel);
        break;
      case 'item.completed':
        currentAssistantBlocks = handleItemCompleted(record, messages, currentAssistantBlocks, sessionModel);
        break;
      case 'turn.completed':
      case 'turn_context':
        flushAssistantBlocks(messages, currentAssistantBlocks, sessionModel);
        currentAssistantBlocks = [];
        break;
    }
  }

  flushAssistantBlocks(messages, currentAssistantBlocks, sessionModel);

  const metadata = extractMetadata(records, messages);
  return buildResult(messages, metadata);
}

function handleEventMsg(record, messages, currentBlocks, sessionModel) {
  const payload = record.payload || {};
  if (payload.type !== 'user_message') return;

  flushAssistantBlocks(messages, currentBlocks, sessionModel);
  const text = payload.message || payload.text || payload.content?.[0]?.text || '';
  if (!text) return;

  messages.push(buildMessage({ role: 'user', contentBlocks: [textBlock(text)], model: sessionModel }));
}

function handleResponseItem(record, messages, currentBlocks, sessionModel) {
  const payload = record.payload || {};
  const role = payload.role;
  const itemType = payload.type;

  switch (itemType) {
    case 'message': {
      if (role === 'assistant') {
        const text = extractResponseItemText(payload);
        if (text) {
          currentBlocks = flushToolCallBlocks(messages, currentBlocks, sessionModel);
          currentBlocks.push(textBlock(text));
        }
      }
      break;
    }
    case 'function_call':
    case 'custom_tool_call': {
      flushAssistantBlocks(messages, currentBlocks, sessionModel);
      currentBlocks = [];
      const name = payload.name || payload.call_id || 'tool_call';
      const input = payload.arguments || payload.input;
      const inputStr = typeof input === 'string' ? truncateStr(input, 200) : truncateStr(JSON.stringify(input), 200);
      currentBlocks.push(toolCallBlock({ name, input: inputStr }));
      flushAssistantBlocks(messages, currentBlocks, sessionModel);
      currentBlocks = [];
      break;
    }
    case 'function_call_output':
    case 'custom_tool_call_output': {
      const output = extractOutputContent(payload);
      const formatted = truncateOutput(output);
      attachOutputToToolCall(messages, formatted);
      break;
    }
    case 'reasoning': {
      const summary = payload.summary?.[0]?.text;
      if (summary) {
        currentBlocks = flushToolCallBlocks(messages, currentBlocks, sessionModel);
        currentBlocks.push(textBlock(summary));
      }
      break;
    }
  }

  return currentBlocks;
}

function handleItemCompleted(record, messages, currentBlocks, sessionModel) {
  const item = record.item || {};
  const itemType = item.type;

  switch (itemType) {
    case 'message': {
      const text = extractMessageText(item);
      if (text) {
        currentBlocks = flushToolCallBlocks(messages, currentBlocks, sessionModel);
        currentBlocks.push(textBlock(text));
      }
      break;
    }
    case 'function_call':
    case 'command': {
      currentBlocks = flushTextBlocks(messages, currentBlocks, sessionModel);
      const name = item.name || item.command || 'command';
      const input = item.arguments || item.input || item.command;
      const rawOutput = item.output || item.result;
      const outputText = unwrapJsonOutput(rawOutput);
      const inputStr = typeof input === 'string' ? input : truncateStr(JSON.stringify(input), 200);
      const outputStr = typeof outputText === 'string' ? outputText : JSON.stringify(outputText);
      currentBlocks.push(toolCallBlock({
        name,
        input: inputStr,
        output: truncateOutput(outputStr),
      }));
      flushAssistantBlocks(messages, currentBlocks, sessionModel);
      currentBlocks = [];
      break;
    }
    case 'file_change': {
      currentBlocks = flushTextBlocks(messages, currentBlocks, sessionModel);
      const path = item.file_path || item.path;
      currentBlocks.push(toolCallBlock({
        name: 'FileChange',
        input: path,
        output: truncateOutput(item.diff || item.content),
      }));
      flushAssistantBlocks(messages, currentBlocks, sessionModel);
      currentBlocks = [];
      break;
    }
  }

  return currentBlocks;
}

function extractResponseItemText(payload) {
  const content = payload.content;
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) {
    return content
      .filter(c => TEXT_CONTENT_TYPES.has(c.type) && c.text)
      .map(c => c.text)
      .join('\n') || null;
  }
  return payload.text || null;
}

function extractMessageText(item) {
  const content = item.content;
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) {
    return content
      .filter(c => c.type === 'text' && c.text)
      .map(c => c.text)
      .join('\n') || null;
  }
  return item.text || null;
}

function extractOutputContent(payload) {
  const raw = payload.output || payload.result;
  return unwrapJsonOutput(raw);
}

function unwrapJsonOutput(raw) {
  if (raw && typeof raw === 'object' && !Array.isArray(raw)) {
    return raw.output || raw.content || JSON.stringify(raw);
  }
  if (typeof raw === 'string') {
    if (raw.trim().startsWith('{')) {
      try {
        const parsed = JSON.parse(raw);
        if (parsed && typeof parsed === 'object') {
          return parsed.output || parsed.content || raw;
        }
      } catch {
        // Not JSON
      }
    }
    return raw;
  }
  return raw != null ? JSON.stringify(raw) : null;
}

function extractSessionModel(records) {
  const sessionMeta = records.find(r => r.type === 'session_meta');
  return sessionMeta?.payload?.model_provider || null;
}

function flushAssistantBlocks(messages, blocks, sessionModel) {
  if (blocks.length === 0) return;
  messages.push(buildMessage({ role: 'assistant', contentBlocks: [...blocks], model: sessionModel }));
  blocks.length = 0;
}

function flushToolCallBlocks(messages, blocks, sessionModel) {
  if (blocks.some(b => b.type === 'tool_call')) {
    flushAssistantBlocks(messages, blocks, sessionModel);
    return [];
  }
  return blocks;
}

function flushTextBlocks(messages, blocks, sessionModel) {
  if (blocks.some(b => b.type === 'text')) {
    flushAssistantBlocks(messages, blocks, sessionModel);
    return [];
  }
  return blocks;
}

function extractMetadata(records, messages) {
  const sessionMeta = records.find(r => r.type === 'session_meta');
  const started = records.find(r => r.type === 'thread.started');

  const metaPayload = sessionMeta?.payload || {};
  const meta = {
    tool_name: 'codex',
    total_messages: messages.length,
  };
  if (metaPayload.id || started?.thread_id) meta.session_id = metaPayload.id || started.thread_id;
  if (metaPayload.timestamp || sessionMeta?.timestamp || started?.timestamp) {
    meta.start_time = metaPayload.timestamp || sessionMeta?.timestamp || started?.timestamp;
  }
  if (metaPayload.model_provider) meta.model = metaPayload.model_provider;
  if (metaPayload.cli_version) meta.cli_version = metaPayload.cli_version;

  return meta;
}
